src/Kvasir/Reconstitution/Lookups.cs

```diff
@@ -30,19 +30,22 @@ internal static partial class Lookup {
         public static object ByKey(DBData key, IEnumerable<object> entities, DataExtractionPlan extractor) {
             Debug.Assert(entities.All(e => e.GetType().IsInstanceOf(extractor.ExpectedSource)));
 
+            if (cacheDictionary_.TryGetValue((key, extractor.ExpectedSource), out object? result)) {
+                return result!;
+            }
+
             foreach (var entity in entities) {
-                var keyIdx = 0;
-                foreach (var value in extractor.ExecutePiecewise(entity)) {
-                    if (!value.Equals(key[keyIdx++])) {
-                        goto nextEntity;
-                    }
+                if (key.SequenceEqual(extractor.ExecutePiecewise(entity!))) {
```

Comment:
If you leave the `!` off this expression (and the one on Line 39), does the compiler throw a warning at you? `entities` is an `IEnumerable<object>`, so the elements _should_ be non-nullable already, which makes the `!` operator unnecessary.

Reply:
You're right - fixed.